### ISSUE TITLE
How to close a JediTerm Widget and it's running OS process correctly?

### ISSUE THREAD

**Issue by manticore-projects** [JetBrains/jediterm#267]
Greetings.

Thank you for providing this great library. We use it to have a full SSH Like Terminal, but via a Web Application (without any SSH Ports open). Simply awesome.

Small problem though: When running a continuous process like `top` in the shell, the process does not get cleaned up properly:

```java
        PtyProcess process = new PtyProcessBuilder().setCommand(command).setEnvironment(envs).start();
        TtyConnector connector = new PtyProcessTtyConnector(process, StandardCharsets.UTF_8);

        JediTermWidget widget = new JediTermWidget(80, 24, new DefaultSettingsProvider());
        widget.setTtyConnector(connector);
        widget.start();

        MDialog dialog = new MDialog(parent, widget, preferredSize, Dialog.ModalityType.MODELESS, "OS Terminal Shell");

        dialog.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                widget.stop();
                widget.close();
                connector.close();
                process.destroyForcibly();
            }

            @Override
            public void windowClosed(WindowEvent e) {
            }
        });
```

I would have expected that the code above closes everything (including the OS process), when closing the window. Without any continuous process it works. But with a continuous process like `top`, a background process keeps running.

So how to close/kill/destroy the OS process correctly please?

**Comment by segrey** [comment]
Sorry for the delay. You probably need to investigate how to properly destroy the underlying process. 

>But with a continuous process like top, a background process keeps running.

Could you please provide more details what you mean that a background process keeps running? For me, `process.destroyForcibly()` sends `SIGKILL` on Unix, so the `top` process is terminated.

**Comment by manticore-projects** [comment]
Let me describe my observation in detail:

1) opening a dialog with Jediterm inside (but without executing any shell command), then closing the dialog works and the whole swing application closes.

2) same as above, but executing a shell command, which terminates/exits (e. g. `ls`, `pwd`, `echo ...`) and then closing the dialog leaves the swing application open (the dialog closes but not all objects get destroyed).

**Comment by segrey** [comment]
Thanks for clarification. Unfortunately, it is not reproduced for me on macOS using the steps:
1. Clone this repo
2. Replace [JediTermMain.kt#L81-L82](https://github.com/JetBrains/jediterm/blob/402dc1c033008f877131be846973ba8c27138b05/JediTerm/src/main/kotlin/com/jediterm/app/JediTermMain.kt#L81-L82) with a single line
```
        arrayOf("top")
```
3. Run the application: `./gradlew :JediTerm:run`. A Swing application is opened with `top` command inside.
4. Close the application frame with mouse. The application is closed as expected.

If it are observing a different behavior, could you please provide a small example project with details how to reproduce it locally?

**Comment by manticore-projects** [comment]
> If it are observing a different behavior, could you please provide a small example project with details how to reproduce it locally?

Sorry for late response, please see below.
When closing **without** calling a OS command, the Application terminates as expected. But when closing **after** calling any OS command (e. g. `ls` or `top`) it does not terminate and stays with a thread running.

```java
package com.manticore.ui;

import com.jediterm.pty.PtyProcessTtyConnector;
import com.jediterm.terminal.TtyConnector;
import com.jediterm.terminal.ui.JediTermWidget;
import com.jediterm.terminal.ui.TerminalWidget;
import com.jediterm.terminal.ui.TerminalWidgetListener;
import com.jediterm.terminal.ui.UIUtil;
import com.jediterm.terminal.ui.settings.DefaultSettingsProvider;
import com.manticore.swingui.MDialog;
import com.pty4j.PtyProcess;
import com.pty4j.PtyProcessBuilder;
import org.junit.jupiter.api.Test;

import javax.swing.*;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

class AdminPanelTest {
    public final static Logger LOGGER = Logger.getLogger(AdminPanelTest.class.getName());

    public static void main(String[] args){
        SwingUtilities.invokeLater(new Runnable() {
            @Override
            public void run() {
                try {
                    Map<String, String> envs = System.getenv();
                    String[] command;
                    if (UIUtil.isWindows) {
                        command = new String[] {"cmd.exe"};
                    } else {
                        command = new String[] {"/bin/sh", "--login"};
                        envs = new HashMap<>(System.getenv());
                        envs.put("TERM", "xterm-256color");
                    }
                    PtyProcess process =
                            new PtyProcessBuilder().setCommand(command).setEnvironment(envs).start();
                    TtyConnector connector = new PtyProcessTtyConnector(process, StandardCharsets.UTF_8);

                    JediTermWidget widget = new JediTermWidget(80, 24, new DefaultSettingsProvider());
                    widget.setTtyConnector(connector);
                    widget.start();

                    JDialog dialog = new JDialog();
                    dialog.add(widget);

                    dialog.addWindowListener(new WindowAdapter() {
                        @Override
                        public void windowClosing(WindowEvent e) {
                            if (widget.isSessionRunning()) {
                                widget.stop();
                                widget.close();
                            }
                            if (connector.isConnected()) {
                                connector.close();
                            }
                            if (process.isAlive()) {
                                process.destroyForcibly();
                            }
                            if (dialog.isShowing()) {
                                dialog.setVisible(false);
                            }
                        }
                    });

                    widget.addListener(new TerminalWidgetListener() {
                        @Override
                        public void allSessionsClosed(TerminalWidget terminalWidget) {
                            if (widget.isSessionRunning()) {
                                widget.stop();
                                widget.close();
                            }
                            if (connector.isConnected()) {
                                connector.close();
                            }
                            if (process.isAlive()) {
                                process.destroyForcibly();
                            }
                            if (dialog.isShowing()) {
                                dialog.setVisible(false);
                            }
                        }
                    });
                    dialog.setDefaultCloseOperation(MDialog.DISPOSE_ON_CLOSE);
                    dialog.setModal(false);
                    dialog.pack();
                    dialog.setVisible(true);
                } catch (Exception ex) {
                    LOGGER.log(Level.SEVERE, ex.getMessage(), ex);
                }
            }
        });
    }
}
```

**Comment by manticore-projects** [comment]
> 4\. Close the application frame with mouse. The application is closed as expected.

I believe, this is achieved by forcefully setting the `Frame.EXIT_ON_CLOSE` which is shading the underlying problem of a rouge thread kept running somewhere.

Of course, I could always call `System.exit()` -- but that is not desirable in a Multi-User/Application Server scenario.

**Comment by segrey** [comment]
@manticore-projects Thanks, reproduced. Fixed by a7670d96cfc65e49f5214750ec5259ca5a9f49eb. Please see the updated [BasicTerminalShellExample.java](https://github.com/JetBrains/jediterm/blob/a7670d96cfc65e49f5214750ec5259ca5a9f49eb/JediTerm/src/main/java/com/jediterm/example/BasicTerminalShellExample.java#L19C14-L19C39). Now JVM should exit gracefully on closing the window with mouse or on exiting the terminal session (`Ctrl+D` or `exit`).

**Comment by manticore-projects** [comment]
Thank you so much for looking into and fixing it. I do appreciate.
Cheers.
